Examples/test-suite/csharp/director_basic_runme.cs

```diff
@@ -69,6 +71,14 @@ class MyOverriddenClass : MyClass {
       throw new Exception("null not received as expected");
     return b;
   }
+
+  public bool nonVirtual() {
+    throw new Exception("non-virtual overrides virtual method");
+  }
+
+  public virtual bool nonOverride() {
```

Comment:
This also is not a valid use case, the C# compiler warns:
```
director_basic/MyClass.cs(96,23): (Location of the symbol related to previous warning)
./director_basic_runme.cs(79,23): warning CS0114: `director_basicNamespace.MyOverriddenClass.nonOverride()' hides inherited member `director_basicNamespace.MyClass.nonOverride()'. To make the current member override that implementation, add the override keyword. Otherwise add the new keyword
director_basic/MyClass.cs(100,23): (Location of the symbol related to previous warning)
```
Similar comments for nonVirtual...
Please adjust the testcase to be valid. Does the patch fix this invalid use case or a valid use case where the `new` and `override` keywords have been added? If it only fixes invalid C# usage, I'll be inclined to reject it as the C# compiler already shows there is a problem. Also, please add code to the director_basic_runme.cs that calls nonOverride() which works now and which failed (got stuck in a recursive loop) before the patch.

Reply:
"invalid usecase" is an overstatement. Compiler does not output an error after all, just a warning. Adding `new` keywords here is a right thing to do, but it only silences warnings and makes code more explicit. In the end compiled result is all the same and presence of `new` keyword does not affect compiled output. In short - patch does make a valid usecase more correct.

Reply:
Warnings are often errors. In this case I don't know what the C# compiler is assuming when neither `new` nor `override` is present. It must be one or the other, do you know? Whichever it is, adding tests that add warnings to the SWIG test-suite is undesirable, so let's make the test explicit and test for both `new` and `override`.

Reply:
When neither `new` nor `override` is present a method in subclass hides a method in parent class. If instance variable is of subclass type it would call method defined in a subclass, however if instance variable is of superclass type it would call method defined in superclass. Indeed having both `new` and `override` does not make much sense and compiler outputs an error.

```cs
class Parent
{
    public virtual int foo1() { return 1; }
    public virtual int foo2() { return 2; }
    public virtual int foo3() { return 3; }
    public virtual int foo4() { return 4; }
}

class Child : Parent
{
    public int foo1() { return 10; }
    public new int foo2() { return 20; }
    public virtual int foo3() { return 30; }
    public override int foo4() { return 40; }
}

var child = new Child();
var parent = (Parent)child;

child.foo1();    // returns 10. Possibly unintentionally hidden method of parent class, produces warning
child.foo2();    // returns 20. Hiding of method in parent class is definitely intended, no warning.
child.foo3();    // returns 30. Does not override method in parent class.
child.foo4();    // returns 40. Does override method in parent class.

parent.foo1();    // returns 1
parent.foo2();    // returns 2
parent.foo3();    // returns 3
parent.foo4();    // returns 40
```

> let's make the test explicit and test for both `new` and `override`.

`new` just removes a warning. Nothing to test here. `override` is already tested by other testcases. This patch deals with method hiding. I believe all cases relevant to the patch are already covered.

Reply:
Thanks, nice detailed explanation. Your change to the testcase is helpful too.